kong/plugins/acl/groups.lua

```diff
@@ -138,9 +125,56 @@ local function get_current_consumer_id()
 end
 
 
+--- Returns a table with all group names.
+-- The table will have an array part to iterate over, and a hash part
+-- where each group name is indexed by itself. Eg.
+-- {
+--   [1] = "users",
+--   [2] = "admins",
+--   users = "users",
+--   admins = "admins",
+-- }
+-- If there are no groups defined, it will return an empty table
+-- @param ctx current ngx.ctx passed as an argument
+-- @return table with groups or nil
+local function get_authenticated_groups()
+  local authenticated_groups = kong.ctx.shared.authenticated_groups
+  if type(authenticated_groups) ~= "table" then
+    authenticated_groups = ngx.ctx.authenticated_groups
+    if type(authenticated_groups) ~= "table" then
+      return nil
+    end
+  end
+
+  local groups = {}
+  for i, group in ipairs(authenticated_groups) do
```

Comment:
Can we use a numeric index instead of ipairs to improve JITability?

Reply:
You mean like this:
```lua
  local groups = {}
  for i = 1, #authenticated_groups do
    local group = authenticated_groups[i]
    groups[i] = group
    groups[group] = group
  end
```

Reply:
Yeah. That extra local probably doesn’t need to be there but it might be optimized away. I’ve found that using ipairs inside of function calls like this results in a new trace generation that causes an abort of the parent trace.